components/hts221/hts221.c

```diff
@@ -0,0 +1,277 @@
+
+    uint8_t status_reg;
+    uint8_t humidity_msb;
+    uint8_t humidity_lsb;
+    uint8_t temperature_msb;
+    uint8_t temperature_lsb;
+    int16_t raw_humidity;
+    int16_t raw_temperature;
+
+    ESP_ERROR_CHECK(hts221_read_status_register(dev, &status_reg));
+    status_reg = status_reg & 0x03;
+
+    if (status_reg != 0x03)
+    {
+        ESP_LOGE(TAG, "No new measurment data is available");
+        return ESP_ERR_INVALID_STATE;
+    }
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, HTS221_REG_HUMIDITY_OUT_L, &humidity_lsb, 1));
+    I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, HTS221_REG_HUMIDITY_OUT_H, &humidity_msb, 1));
+    I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, HTS221_REG_TEMP_OUT_L, &temperature_lsb, 1));
+    I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, HTS221_REG_TEMP_OUT_H, &temperature_msb, 1));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    raw_humidity = (humidity_msb << 8) | humidity_lsb;
+    raw_temperature = (temperature_msb << 8) | temperature_lsb;
+
+    *humidity = (float)(calibration_parameters.H1_rH - calibration_parameters.H0_rH) *
+                    (raw_humidity - calibration_parameters.H0_T0_OUT) /
+                    (float)(calibration_parameters.H1_T0_OUT - calibration_parameters.H0_T0_OUT) +
+                (float)calibration_parameters.H0_rH;
+    if (*humidity > 100)
+    {
+        *humidity = 100;
+    }
+    *temperature = (float)(calibration_parameters.T1_degC - calibration_parameters.T0_degC) *
+                       (raw_temperature - calibration_parameters.T0_OUT) /
+                       (float)(calibration_parameters.T1_OUT - calibration_parameters.T0_OUT) +
+                   (float)calibration_parameters.T0_degC;
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_who_am_i(i2c_dev_t *dev, uint8_t *who_am_i)
+{
+    CHECK_ARG(dev && who_am_i);
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, HTS221_REG_WHOAMI, who_am_i, 1));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_read_av_conf(i2c_dev_t *dev, hts221_temperature_avg_t *t_avg, hts221_humidity_avg_t *rh_avg)
+{
+    CHECK_ARG(dev && t_avg && rh_avg);
+
+    uint8_t av_conf;
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, HTS221_REG_AV_CONF, &av_conf, 1));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    *t_avg = (hts221_temperature_avg_t)((av_conf & 0x38) >> 3);
+    *rh_avg = (hts221_humidity_avg_t)((av_conf & 0x07));
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_set_av_conf(i2c_dev_t *dev, hts221_temperature_avg_t *t_avg, hts221_humidity_avg_t *rh_avg)
+{
+    CHECK_ARG(dev && t_avg && rh_avg);
+
+    uint8_t av_conf;
+
+    av_conf = (uint8_t)((*t_avg) << 3) | (*rh_avg);
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_write_reg(dev, HTS221_REG_AV_CONF, &av_conf, 1));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_power_toggle(i2c_dev_t *dev, hts221_power_t *power)
```

Comment:
Use bool here instead of `hts221_power_t`

Reply:
also done